src/groups/mqb/mqbblp/mqbblp_queuehandle.cpp

```diff
@@ -1173,22 +1249,78 @@ void QueueHandle::onAckMessage(const bmqp::AckMessage& ackMessage)
     // was requested or queue failed to post the message or both.  Also note
     // that we need to reset the queueId in ack message to the one which is
     // known downstream.
-    mqbi::DispatcherClient* client = d_clientContext_sp->client();
-    mqbi::DispatcherEvent*  event  = client->dispatcher()->getEvent(client);
-    (*event)
-        .setType(mqbi::DispatcherEventType::e_ACK)
-        .setSource(d_queue_sp.get())
-        .setAckMessage(ackMessage);
 
-    // Override with correct downstream queueId
-    const mqbi::DispatcherAckEvent* ackEvent = event->asAckEvent();
+    d_domainStats_p->onEvent(mqbstat::QueueStatsDomain::EventType::e_ACK, 1);
 
-    bmqp::AckMessage& ackMsg = const_cast<bmqp::AckMessage&>(
-        ackEvent->ackMessage());
-    ackMsg.setQueueId(id());
+    mqbi::InlineClient* inlineClient = d_clientContext_sp->inlineClient();
 
-    client->dispatcher()->dispatchEvent(event, client);
-    d_domainStats_p->onEvent(mqbstat::QueueStatsDomain::EventType::e_ACK, 1);
+    if (inlineClient == 0) {
+        mqbi::DispatcherClient* client = d_clientContext_sp->client();
+        mqbi::DispatcherEvent*  event = client->dispatcher()->getEvent(client);
+        (*event)
+            .setType(mqbi::DispatcherEventType::e_ACK)
+            .setSource(d_queue_sp.get())
+            .setAckMessage(ackMessage);
+
+        // Override with correct downstream queueId
+        const mqbi::DispatcherAckEvent* ackEvent = event->asAckEvent();
+
+        bmqp::AckMessage& ackMsg = const_cast<bmqp::AckMessage&>(
+            ackEvent->ackMessage());
+        ackMsg.setQueueId(id());
+
+        client->dispatcher()->dispatchEvent(event, client);
+        return;  // RETURN
+    }
+
+    if (BSLS_PERFORMANCEHINT_PREDICT_UNLIKELY(status !=
+                                              bmqt::AckResult::e_SUCCESS)) {
+        BSLS_PERFORMANCEHINT_UNLIKELY_HINT;
+
+        // Throttle error log if this is a 'failed Ack': note that we log at
+        // INFO level in order not to overwhelm the dashboard, if a queue is
+        // full, every post will nack, which could be a lot.
+        if (d_throttledFailedAckMessages.requestPermission()) {
+            BALL_LOG_INFO << "Queue '" << d_queue_sp->description()
+                          << "' NACK "
+                          << "[status: " << status
+                          << ", GUID: " << ackMessage.messageGUID()
+                          << ", node '" << d_clientContext_sp->description()
+                          << "']";
```

Comment:
This log line is logged only when we don't have an inline client, is it okay? The same is true for other logs later in this method.